### PR TITLE
Use the right type on Response for tests

### PR DESCRIPTION
**Pull request by Kludex**
## Changes
- Replace `starlette.responses.Response` by `requests.Response` on tests, as it's the right type returned by `TestClient`.
- Replace middleware implementation from `BaseHTTPMiddleware` to pure ASGI.

Can you approve the pipeline @trallnag ? cc @tiangolo - Some tests still failing atm.

### CODE REVIEW

File: tests/test_instrumentation.py [hunk 1]
```diff
@@ -100,7 +101,7 @@ def get_response(client: TestClient, path: str) -> Response:
     return response
 
 
-def assert_is_not_multiprocess(response: Response) -> None:
+def assert_is_not_multiprocess(response: TestClientResponse) -> None:
     assert response.status_code == 200
     assert b"Multiprocess" not in response.content
     assert b"# HELP process_cpu_seconds_total" in response.content
```
Comment (Kludex) [inline tests/test_instrumentation.py:107]: This line of code doesn't pass locally... Which is incredibly annoying.

**Comment by trallnag** [comment]
Hey @Kludex, thanks for your contribution efforts. Generally I'm open towards switching to a pure ASGI implementation. But sadly I can't support you with the implementation because I lack knowledge regarding ASGI and async Python in general. Can you tell me what the advantage of switching to an ASGI implementation?

**Comment by trallnag** [comment]
Some tests are failing. It seems like something about the way paths / handlers are exposed has changed. For example this test is failing. I think it is due to the instrumentator not passing all parameters to the middleware. Looking at `excluded_handlers`

File: prometheus_fastapi_instrumentator/instrumentation.py [hunk 2]
```diff
@@ -119,78 +119,20 @@ def instrument(self, app: FastAPI):
...
+            should_ignore_untemplated=self.should_ignore_untemplated,
+            should_group_untemplated=self.should_group_untemplated,
+            should_round_latency_decimals=self.should_round_latency_decimals,
+            should_respect_env_var=self.should_respect_env_var,
+            should_instrument_requests_inprogress=self.should_instrument_requests_inprogress,
+            round_latency_decimals=self.round_latency_decimals,
+            env_var_name=self.env_var_name,
+            inprogress_name=self.inprogress_name,
+            inprogress_labels=self.inprogress_labels,
+            instrumentations=self.instrumentations,
```
Comment (adriangb) [inline prometheus_fastapi_instrumentator/instrumentation.py:134]:
```suggestion
            instrumentations=self.instrumentations,
            excluded_handlers=self.excluded_handlers,
```

This fixes the tests it seems

**Review by adriangb — CHANGES_REQUESTED** [review]

**Comment by adriangb** [comment]
> Can you tell me what the advantage of switching to an ASGI implementation?

`BaseHTTPMiddleware` (which is what `app.middleware` uses) is a wrapper around generic ASGI middleware to expose a request/response interface. Internally it is _very_ complex, and has several bugs which we don't think can ever be resolved. Because of this, we discourage its use. You can read more here: https://www.starlette.io/middleware/#basehttpmiddleware

Unrelated this this PR: just FYI I couldn't get tests to run on an M1 Mac, I had to run them in a Docker container.

**Comment by Kludex** [comment]
Thanks 👍

**Comment by Kludex** [comment]
I didn't notice the title was wrong 😅

**Comment by tiangolo** [comment]
Niceee! Thanks for this! 🙇

**Pull request merged**